quickjs.c

```diff
@@ -14574,6 +14574,7 @@ static JSValue JS_CallInternal(JSContext *caller_ctx, JSValue func_obj,
     if (js_check_stack_overflow(rt, alloca_size))
         return JS_ThrowStackOverflow(caller_ctx);
 
+    sf->cur_pc = NULL;
```

Comment:
This is not the correct fix: you should instead initialize `sf->cur_pc` along with `pc` on line 14603, before the stack frame gets linked.

See the explanation in the issue conversation #323

Reply:
Will do!

Reply:
You mean to initialize it to NULL?

Reply:
No, not to `NULL`:
`` `c
    pc = b->byte_code_buf;
    s->cur_pc = pc;
```
This allows `build_backtrace()` can find the source position for the code that ultimately caused the error to be thrown even if it is triggered by an opcode that did not set `sf->cur` prior to calling some function in `JS_CallInternal()`.

Reply:
Many more opcodes need to set `sf->cur_pc`... almost all of those that test for exceptions and `goto exception`:

- `OP_fclosure8`
- `OP_get_length`
- `OP_push_this` before the call to `JS_ToObject`
- `OP_object`, `OP_special_object`
- `OP_rest`
- `OP_fclosure`
- `OP_array_from`
- `OP_check_ctor_return`, `OP_check_ctor`
- `OP_check_brand`, `OP_add_brand`
- `OP_throw` (if even `JS_Throw()` gets instrumented for a debugger)
- `OP_throw_error`
- `OP_apply_eval`
- `OP_regexp`
- `OP_get_super`
- `OP_import`
- `OP_check_var`
- `OP_get_var_undef`, `OP_get_var`, `OP_put_var`, `OP_put_var_init`, `OP_put_var_strict`
- `OP_check_define_var`, `OP_define_var`, `OP_define_func`
- `OP_get_var_ref_check`, `OP_put_var_ref_check`, `OP_put_var_ref_check_init`
- `OP_get_loc_check` and `OP_put_loc_check` before the call to `JS_ThrowReferenceErrorUninitialized2`
- `OP_put_loc_check_init` before the call to `JS_ThrowReferenceError`
- `OP_make_loc_ref`, `OP_make_arg_ref`, `OP_make_var_ref_ref`, `OP_make_var_ref`
- `OP_if_false` and `OP_if_true` and the short variants before the call to `js_poll_interrupts` (and all such opcodes)
- `OP_ret` before the call to `JS_ThrowInternalError`
- `OP_for_in_start`, `OP_for_in_next`, `OP_for_of_start`, `OP_for_of_next`...
- etc.

Reply:
Got it!

Reply:
Better set both `pc` and `sf->cur_pc` at the same spot IMHO

Reply:
Sure. Before the opcode switch the frame is "initialized". The pc is known at that point, but we then might change it, that's why I wondered if setting cur_pc before the dispatch could have adverse affects.

Reply:
> I wondered if setting cur_pc before the dispatch could have adverse affects.

I don't think so.  The stack frame is just uninitialized there and it is a potential problem.
The alternate path where the generator function is restarted gets `pc` from `sf->cur_pc` and branches to `restart:`, so no conflict there either.

Reply:
Alright. So then it looks like we could get away with doing that instead of setting it in each opcode. I'll dig some more tonight!

Reply:
No, initializing it before the switch is a safety measure to avoid a potential crash if `sf->cur_pc` gets used without having been set, but it is not sufficient for `build_backtrace()` to get the correct source location when the stack trace is built.
We use `pc` instead of `sf->cur_pc` for the dispatch to optimize the code and let the compiler use a register instead of updating the stack frame for each opcode. Same for `sp` and `sf->cur_sp`.

Reply:
I see. In that case it might be better to initialize it to NULL and add an asset inside build_backtrace so we can catch 'em all. WDYT?

Reply:
We could do that.  `sf->cur_pc` should not be null for bytecoded functions, but is null for C functions.